utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala

```diff
@@ -0,0 +1,214 @@
+/**
+ * Licensed to Big Data Genomics (BDG) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The BDG licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.utils.rangearray
+
+import com.esotericsoftware.kryo.io.{ Input, Output }
+import com.esotericsoftware.kryo.{ Kryo, Serializer }
+import org.apache.spark.rdd.RDD
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+/**
+ * Companion object for building a RangeSearchableArray from an RDD.
+ */
+object RangeSearchableArray extends Serializable {
+
+  /**
+   * Sorts the RDD and collects it to build RangeSearchableArray, a sorted array that searches
+   * over ranges. This is used for a left side of the broadcast region join in ADAM.
+   *
+   * @param rdd RDD to build a RangeSearchableArray from.
+   * @return The RangeSearchableArray built from this RDD.
+   */
+  def apply[K <: Interval[K]: ClassTag, T: ClassTag](rdd: RDD[(K, T)]): RangeSearchableArray[K, T] = {
+    val sortedArray =
+      rdd.sortByKey()
+        .collect
+
+    new RangeSearchableArray(sortedArray, sorted = true)
+  }
+}
+
+/**
+ * Originally, a RangeSearchableArray was a collection of trees.
+ * Alas, we have no trees anymore.
+ * I blame global warming.
+ *
+ * @param arr An array of values for the left side of the join. We require
+ *   this array to be sorted.
+ */
+case class RangeSearchableArray[K <: Interval[K], T: ClassTag](arr: Array[(K, T)], sorted: Boolean = false) {
+
+  // ensure that array is sorted
+  val array =
```

Comment:
This looks new. Is this for the append case? If so, let's make a few changes:

* Make `val array` private
* Make `RangeSearchableArray`:
  * A class and not a case class
  * extend Serializable